server/src/main/java/org/candlepin/controller/OwnerProductShareManager.java

```diff
@@ -0,0 +1,247 @@
+
+    /**
+     * Represents a resolved product post sharing
+     */
+    public static class ResolvedProduct {
+
+        private Product product;
+        /**
+         * true if there is a refresh due for this product in the recipient, due to recent change in share
+         * resolution.
+         */
+        private boolean refreshDue;
+
+        public ResolvedProduct(Product product) {
+            this.product = product;
+        }
+
+        public ResolvedProduct(Product product, boolean refreshDue) {
+            this.product = product;
+            this.refreshDue = refreshDue;
+        }
+
+        public Product getProduct() {
+            return product;
+        }
+
+        public void setProduct(Product product) {
+            this.product = product;
+        }
+
+        public boolean isRefreshDue() {
+            return refreshDue;
+        }
+
+        public void setRefreshDue(boolean refreshDue) {
+            this.refreshDue = refreshDue;
+        }
+    }
+
+    @Inject
+    public OwnerProductShareManager(OwnerProductCurator ownerProductCurator,
+        OwnerProductShareCurator shareCurator) {
+        this.ownerProductCurator = ownerProductCurator;
+        this.shareCurator = shareCurator;
+    }
+
+    /**
+     *
+     * @param recipientOwner the owner whose resolved products are being requested
+     * @param productIds the id of the products being requested
+     * @param includeShares whether to include shared products or strictly return only manifest imported or
+     *                      upstream products
+     * @return recipient's ownerProduct if they exist, or resolve the latest shared product from shared
+     * products if shares are requested to be included.
+     */
+    public Product resolveProductById(Owner recipientOwner, String productId, boolean includeShares) {
+        List<Product> products = resolveProductsByIds(recipientOwner,
+            Collections.singleton(productId), includeShares);
+        if (CollectionUtils.isNotEmpty(products)) {
+            return products.get(0);
+        }
+        return null;
+    }
+
+    /**
+     *
+     * @param recipientOwner the owner whose resolved product is being requested
+     * @param product the id of the product being requested
+     * @param includeShares whether to include shared products or strictly return only manifest imported or
+     *                      upstream product
+     * @return recipient's ownerProduct if it exists, or resolve the most recently shared product from shared
+     * products if shares are requested to be included.
+     */
+    public List<Product> resolveProductsByIds(Owner recipientOwner, Collection<String> productIds,
+        boolean includeShares) {
+        Map<String, Product> result = new HashMap<String, Product>();
+        List<Product> existingProducts;
+        Set<String> missingProducts = new HashSet<String>();
+
+        if (productIds != null && !productIds.isEmpty()) {
```

Comment:
I'm still confused why `productIds` would ever be null or empty.  If that's an expected case, I think we should have a different public method named something like `resolveAllProducts(Owner owner, boolean includeShares)` and then have it and `resolveProductsByIds` delegate to the same private method.

Reply:
note from discussion: move productid logic to resource